File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-mktenv-5-4/TermCurve.cpp
// TermCurve.cpp 
#include "TermCurve.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

TermCurve::TermCurve(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //interpolationMethodNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* interpolationMethodNode = xmlNode->FirstChildElement("interpolationMethod");

   if(interpolationMethodNode){interpolationMethodIsNull_ = false;}
   else{interpolationMethodIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- interpolationMethodNode , address : " << interpolationMethodNode << std::endl;
   #endif
   if(interpolationMethodNode)
   {
      if(interpolationMethodNode->Attribute("href") || interpolationMethodNode->Attribute("id"))
      {
          if(interpolationMethodNode->Attribute("id"))
          {
             interpolationMethodIDRef_ = interpolationMethodNode->Attribute("id");
             interpolationMethod_ = boost::shared_ptr<InterpolationMethod>(new InterpolationMethod(interpolationMethodNode));
             interpolationMethod_->setID(interpolationMethodIDRef_);
             IDManager::instance().setID(interpolationMethodIDRef_,interpolationMethod_);
          }
          else if(interpolationMethodNode->Attribute("href")) { interpolationMethodIDRef_ = interpolationMethodNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { interpolationMethod_ = boost::shared_ptr<InterpolationMethod>(new InterpolationMethod(interpolationMethodNode));}
   }

   //extrapolationPermittedNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* extrapolationPermittedNode = xmlNode->FirstChildElement("extrapolationPermitted");

   if(extrapolationPermittedNode){extrapolationPermittedIsNull_ = false;}
   else{extrapolationPermittedIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- extrapolationPermittedNode , address : " << extrapolationPermittedNode << std::endl;
   #endif
   if(extrapolationPermittedNode)
   {
      if(extrapolationPermittedNode->Attribute("href") || extrapolationPermittedNode->Attribute("id"))
      {
          if(extrapolationPermittedNode->Attribute("id"))
          {
             extrapolationPermittedIDRef_ = extrapolationPermittedNode->Attribute("id");
             extrapolationPermitted_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(extrapolationPermittedNode));
             extrapolationPermitted_->setID(extrapolationPermittedIDRef_);
             IDManager::instance().setID(extrapolationPermittedIDRef_,extrapolationPermitted_);
          }
          else if(extrapolationPermittedNode->Attribute("href")) { extrapolationPermittedIDRef_ = extrapolationPermittedNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { extrapolationPermitted_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(extrapolationPermittedNode));}
   }

   //pointNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* pointNode = xmlNode->FirstChildElement("point");

   if(pointNode){pointIsNull_ = false;}
   else{pointIsNull_ = true;}

   if(pointNode)
   {
      for(pointNode; pointNode; pointNode = pointNode->NextSiblingElement("point")){
          #ifdef ConsolePrint
              FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- pointNode , address : " << pointNode << std::endl;
          #endif
          if(pointNode->Attribute("href") || pointNode->Attribute("id"))
          {
              if(pointNode->Attribute("id"))
              {
                  pointIDRef_ = pointNode->Attribute("id");
                  point_.push_back(boost::shared_ptr<TermPoint>(new TermPoint(pointNode)));
                  point_.back()->setID(pointIDRef_);
                  IDManager::instance().setID(pointIDRef_, point_.back());
              }
              else if(pointNode->Attribute("href")) { pointIDRef_ = pointNode->Attribute("href");}
              else { QL_FAIL("id or href error"); }
          }
          else { point_.push_back(boost::shared_ptr<TermPoint>(new TermPoint(pointNode)));}
      }
   }
   else {
       #ifdef ConsolePrint
           FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- pointNode , address : " << pointNode << std::endl;
       #endif
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<InterpolationMethod> TermCurve::getInterpolationMethod()
{
   if(!this->interpolationMethodIsNull_){
        if(interpolationMethodIDRef_ != ""){
             return boost::shared_static_cast<InterpolationMethod>(IDManager::instance().getID(interpolationMethodIDRef_));
        }else{
             return this->interpolationMethod_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<InterpolationMethod>();
   }
}
boost::shared_ptr<XsdTypeBoolean> TermCurve::getExtrapolationPermitted()
{
   if(!this->extrapolationPermittedIsNull_){
        if(extrapolationPermittedIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeBoolean>(IDManager::instance().getID(extrapolationPermittedIDRef_));
        }else{
             return this->extrapolationPermitted_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeBoolean>();
   }
}
std::vector<boost::shared_ptr<TermPoint>> TermCurve::getPoint()
{
   if(!this->pointIsNull_){
        if(pointIDRef_ != ""){
             return this->point_;
        }else{
             return this->point_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return std::vector<boost::shared_ptr<TermPoint>>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-mktenv-5-4/TermCurve.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class TermCurve : ISerialized
    {
        public TermCurve(XmlNode xmlNode)
        {
            XmlNode interpolationMethodNode = xmlNode.SelectSingleNode("interpolationMethod");
            
            if (interpolationMethodNode != null)
            {
                if (interpolationMethodNode.Attributes["href"] != null || interpolationMethodNode.Attributes["id"] != null) 
                {
                    if (interpolationMethodNode.Attributes["id"] != null) 
                    {
                        interpolationMethodIDRef_ = interpolationMethodNode.Attributes["id"].Value;
                        InterpolationMethod ob = new InterpolationMethod(interpolationMethodNode);
                        IDManager.SetID(interpolationMethodIDRef_, ob);
                    }
                    else if (interpolationMethodNode.Attributes["href"] != null)
                    {
                        interpolationMethodIDRef_ = interpolationMethodNode.Attributes["href"].Value;
                    }
                    else
                    {
                        interpolationMethod_ = new InterpolationMethod(interpolationMethodNode);
                    }
                }
                else
                {
                    interpolationMethod_ = new InterpolationMethod(interpolationMethodNode);
                }
            }
            
        
            XmlNode extrapolationPermittedNode = xmlNode.SelectSingleNode("extrapolationPermitted");
            
            if (extrapolationPermittedNode != null)
            {
                if (extrapolationPermittedNode.Attributes["href"] != null || extrapolationPermittedNode.Attributes["id"] != null) 
                {
                    if (extrapolationPermittedNode.Attributes["id"] != null) 
                    {
                        extrapolationPermittedIDRef_ = extrapolationPermittedNode.Attributes["id"].Value;
                        XsdTypeBoolean ob = new XsdTypeBoolean(extrapolationPermittedNode);
                        IDManager.SetID(extrapolationPermittedIDRef_, ob);
                    }
                    else if (extrapolationPermittedNode.Attributes["href"] != null)
                    {
                        extrapolationPermittedIDRef_ = extrapolationPermittedNode.Attributes["href"].Value;
                    }
                    else
                    {
                        extrapolationPermitted_ = new XsdTypeBoolean(extrapolationPermittedNode);
                    }
                }
                else
                {
                    extrapolationPermitted_ = new XsdTypeBoolean(extrapolationPermittedNode);
                }
            }
            
        
            XmlNodeList pointNodeList = xmlNode.SelectNodes("point");
            
            if (pointNodeList != null)
            {
                this.point_ = new List<TermPoint>();
                foreach (XmlNode item in pointNodeList)
                {
                    if (item.Attributes["href"] != null || item.Attributes["id"] != null) 
                    {
                        if (item.Attributes["id"] != null) 
                        {
                            pointIDRef_ = item.Attributes["id"].Value;
                            point_.Add(new TermPoint(item));
                            IDManager.SetID(pointIDRef_, point_[point_.Count - 1 ]);
                        }
                        else if (item.Attributes["href"] != null)
                        {
                            pointIDRef_ = item.Attributes["href"].Value;
                        }
                        else
                        {
                        point_.Add(new TermPoint(item));
                        }
                    }
                    else
                    {
                        point_.Add(new TermPoint(item));
                    }
                }
            }
            
        
        }
        
    
        #region InterpolationMethod_
        private InterpolationMethod interpolationMethod_;
        public InterpolationMethod InterpolationMethod_
        {
            get
            {
                if (this.interpolationMethod_ != null)
                {
                    return this.interpolationMethod_; 
                }
                else if (this.interpolationMethodIDRef_ != null)
                {
                    interpolationMethod_ = IDManager.getID(interpolationMethodIDRef_) as InterpolationMethod;
                    return this.interpolationMethod_; 
                }
                else
                {
                      return this.interpolationMethod_; 
                }
            }
            set
            {
                if (this.interpolationMethod_ != value)
                {
                    this.interpolationMethod_ = value;
                }
            }
        }
        #endregion
        
        public string interpolationMethodIDRef_ { get; set; }
        #region ExtrapolationPermitted_
        private XsdTypeBoolean extrapolationPermitted_;
        public XsdTypeBoolean ExtrapolationPermitted_
        {
            get
            {
                if (this.extrapolationPermitted_ != null)
                {
                    return this.extrapolationPermitted_; 
                }
                else if (this.extrapolationPermittedIDRef_ != null)
                {
                    extrapolationPermitted_ = IDManager.getID(extrapolationPermittedIDRef_) as XsdTypeBoolean;
                    return this.extrapolationPermitted_; 
                }
                else
                {
                      return this.extrapolationPermitted_; 
                }
            }
            set
            {
                if (this.extrapolationPermitted_ != value)
                {
                    this.extrapolationPermitted_ = value;
                }
            }
        }
        #endregion
        
        public string extrapolationPermittedIDRef_ { get; set; }
        #region Point_
        private List<TermPoint> point_;
        public List<TermPoint> Point_
        {
            get
            {
                if (this.point_ != null)
                {
                    return this.point_; 
                }
                else if (this.pointIDRef_ != null)
                {
                    return this.point_; 
                }
                else
                {
                      return this.point_; 
                }
            }
            set
            {
                if (this.point_ != value)
                {
                    this.point_ = value;
                }
            }
        }
        #endregion
        
        public string pointIDRef_ { get; set; }
        
    
        
    
    }
    
}

